tycho-core/src/main/java/org/eclipse/tycho/p2resolver/TargetDefinitionVariableResolver.java

```diff
@@ -0,0 +1,99 @@
+package org.eclipse.tycho.p2resolver;
+
+import java.util.function.Function;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.codehaus.plexus.component.annotations.Component;
+import org.codehaus.plexus.component.annotations.Requirement;
+import org.codehaus.plexus.personality.plexus.lifecycle.phase.Initializable;
+import org.codehaus.plexus.personality.plexus.lifecycle.phase.InitializationException;
+import org.eclipse.tycho.ReactorProject;
+import org.eclipse.tycho.core.shared.MavenContext;
+import org.eclipse.tycho.core.shared.MavenLogger;
+
+@Component(role = TargetDefinitionVariableResolver.class)
+public class TargetDefinitionVariableResolver implements Initializable {
+
+    private static final Pattern SYSTEM_PROPERTY_PATTERN = createVariablePatternArgument("system_property");
+    private static final Pattern PROJECT_LOC_PATTERN = createVariablePatternArgument("project_loc");
+    private static final Pattern ENV_VAR_PATTERN = createVariablePatternArgument("env_var");
+
+    @Requirement
+    private MavenContext mavenContext;
+    private MavenLogger logger;
```

Comment:
Please use `org.codehaus.plexus.logging.Logger` like this
```suggestion
    @Requirement
    private Logger logger;
```

that way we don't need the `Initializable` nor the `MavenContext`

Reply:
I replaced the logger as you suggest, but we still do need the MavenContext. It was not used just to get to the logger, I also need it to `mavenContext.getProjects()` and `mavenContext.getSessionProperties()`. What are suitable replacements for these?

Reply:
You can take a look at `DefaultMavenContext` class,

mavenContext.getSessionProperties() --> `Session.getProperties()`
mavenContext.getProjects() --> `Session.getProjects()`

but if you like you can just retain this for now and only replace the logger, its just good to remove references to the MavenContext where it is easily possible (like loggers).

Reply:
I looked there but if I understand you correctly that would mean instead of `MavenContext` I'd have to inject `LegacySupport` and copy/paste the `get(Session|Propeties|Projects)` methods from `DefaultMavenContext` into the `TargetDefinitionVariableResolver` which hardly seems as an improvement

Reply:
If you want access to the session, you can add `@SessionScoped` to the component and add a constructor like this:

```
@Inject
public MySessionScopedComponent(MavenSession mavenSession) {
    this.mavenSession = mavenSession;
}
```

Reply:
Ok I did that but now there is a new problem... 
In unit tests I need to provide some `TargetDefinitionVariableResolver`. 
Before I just passed in real impl with mocked `MavenContext` I already had available.
Now I'd need to provide mocked `MavenSession` which seems like a lot of trouble, so I opted to rather turn `TargetDefinitionVariableResolver` into an interface and have its implementation
``` java
@SessionScoped
@Component(role = TargetDefinitionVariableResolver.class)
public class DefaultTargetDefinitionVariableResolver implements TargetDefinitionVariableResolver {
    @Requirement
    private Logger logger;

    @Inject
    public DefaultTargetDefinitionVariableResolver(MavenSession session) {
        this.mavenSession = session;
    }
```
Now this works great in the unit tests, but it started causing trouble in tests inherited from `AbstractTychoMojoTestCase` as the real impl with injeted session is used here (any way to place mocks into the plexus container in these tests?) and apparently the `MavenSession` that gets injected into it is different than the session prepared inside `AbstractTychoMojoTestCase#setUp`. In particular I tracked it down to this discrepancy - when the failure of the test occurs there is stack trace like this
```
DefaultTargetDefinitionVariableResolver.projects() line: 102    
DefaultTargetDefinitionVariableResolver.findProjectLocation(String) line: 70    
0x0000000801817800.apply(Object) line: not available    
DefaultTargetDefinitionVariableResolver.resolvePattern(String, Pattern, Function<String,String>) line: 110  
DefaultTargetDefinitionVariableResolver.resolve(String) line: 61    
TargetDefinitionResolver.resolvePath(String, TargetDefinition) line: 294    
TargetDefinitionResolver.resolveContentWithExceptions(TargetDefinition, IProvisioningAgent) line: 146   
TargetDefinitionResolver.resolveContent(TargetDefinition, IProvisioningAgent) line: 102 
TargetDefinitionResolverService.resolveFromArguments(ResolutionArguments) line: 93  
0x00000008018131d0.apply(Object) line: not available    
ConcurrentHashMap<K,V>.computeIfAbsent(K, Function<? super K,? extends V>) line: 1708   
TargetDefinitionResolverService.getTargetDefinitionContent(TargetDefinition, List<TargetEnvironment>, ExecutionEnvironmentResolutionHints, IncludeSourceMode, IProvisioningAgent) line: 68  
TargetPlatformFactoryImpl.resolveTargetDefinitions(TargetPlatformConfigurationStub, ExecutionEnvironmentResolutionHints) line: 214  
TargetPlatformFactoryImpl.createTargetPlatform(TargetPlatformConfigurationStub, ExecutionEnvironmentResolutionHandler, List<ReactorProject>) line: 164  
TargetPlatformFactoryImpl.createTargetPlatform(TargetPlatformConfigurationStub, ExecutionEnvironmentConfiguration, List<ReactorProject>) line: 137  
TargetPlatformFactoryImpl.createTargetPlatform(TargetPlatformConfigurationStub, ExecutionEnvironmentConfiguration, List) line: 88   
ReactorRepositoryManagerImpl.computePreliminaryTargetPlatform(ReactorProject, TargetPlatformConfigurationStub, ExecutionEnvironmentConfiguration, List<ReactorProject>) line: 71    
P2DependencyResolver.computePreliminaryTargetPlatform(MavenSession, MavenProject, List<ReactorProject>) line: 205   
...
```
and in `P2DependencyResolver` its `MavenSession#getProjects` returns
```
MavenProject: bootclasspath:bootclasspath:0.0.1 @ /data/codasip/git/vaclavHala_tycho/tycho-core/target/projects/bootclasspath/pom.xml,
MavenProject: bootclasspath:bundle1:0.0.1 @ /data/codasip/git/vaclavHala_tycho/tycho-core/target/projects/bootclasspath/bundle1/pom.xml,
MavenProject: bootclasspath:bundle2:0.0.1 @ /data/codasip/git/vaclavHala_tycho/tycho-core/target/projects/bootclasspath/bundle2/pom.xml
```
whereas in `DefaultTargetDefinitionVariableResolver` its `MavenSession#getProjects` returns
```
MavenProject: unknown:empty-project:0 @ 
```

**TL;DR**
I'd prefer to stick with `MavenContext` for now as replacing it by `MavenSession` seems to have far reaching consequences I don't fully understand and don't feel able to properly handle

Reply:
Just make sure the project is a `TychoPlexusTestCase` then you can simply `lookup(TargetDefinitionVariableResolver.class)`

Reply:
Sorry I can't see how that helps me, all the failing tests already implement `PlexusTestCase` which probides the `lookup` method. Problem is not how to get to `TargetDefinitionVariableResolver`, but how to setup the test such that anyone who looks up the resolver using the method you show will get some mock implementation instead of the real one. I have pushed commit showing the problem

Reply:
You should never need to use any mocks to test your changes, only test the user visible API and/or using real integration test projects.

Reply:
Beside that, if you want to change the session values in a test you can override `org.eclipse.tycho.testing.TychoPlexusTestCase.modifySession(MavenSession)`

Reply:
Regarding failures like:

`Caused by: java.lang.NullPointerException: Cannot invoke "java.io.File.getName()" because the return value of "org.eclipse.tycho.ReactorProject.getBasedir()" is null`

one should simply add a null-check at that location.

Reply:
> one should simply add a null-check at that location.

I don't think I agree here, these tests worked fine before but started breaking when I changed `MavenContext` in the variable resolver for `MavenSession` to look up current projects. The `MavenSession` that gets injected provides different list of projects than `MavenContext`. Projects provided by the `MavenSession` (as shown in my comment above) do not have the `basedir` set which causes these NPE failures.

I agree mocking is probably not the way to do here. However simply injecting the `MavenSession` leads to the test failures for reasons I do not currently understand and don't think I could address in reasonable time. So since using `MavenSession` instead of `MavenContext` is not a matter of simple 1:1 replacement I suggest we do not go through with it in this PR and keep the `MavenContext` for now

Reply:
> So since using MavenSession instead of MavenContext is not a matter of simple 1:1 replacement I suggest we do not go through with it in this PR and keep the MavenContext for now

That would be okay.